// vidstack/dist/prod/define/vidstack-audio-layout.js
import { x, n } from '../chunks/vidstack-Bf6opHxk.js';
import { n as signal, e as effect, Q as toggleClass, H as Host, o as onDispose, l as listenEvent, s as setAttribute } from '../chunks/vidstack-DoOmgEfT.js';
import { D as DefaultLayout, a as DefaultSeekButton, b as DefaultPlayButton, c as DefaultTimeSlider, d as DefaultTimeInvert, e as DefaultCaptionButton, f as DefaultChaptersMenu, g as DefaultSettingsMenu, u as useDefaultLayoutContext, h as DefaultControlsSpacer, i as DefaultMuteButton, j as DefaultVolumeSlider, k as i18n, l as DefaultChapterTitle, m as createMenuContainer, n as DefaultLayoutIconsLoader } from '../chunks/vidstack-DB3I3nZp.js';
import { u as useMediaContext, b as useMediaState } from '../chunks/vidstack-i0oJKwBM.js';
import { $ as $signal, S as SlotManager } from '../chunks/vidstack-B2Ww2rqm.js';
import { L as LitElement } from '../chunks/vidstack-CbhL8rpf.js';
import { u as useTransitionActive, b as useResizeObserver, d as useActive, e as useMouseEnter, f as useRectCSSVars } from '../chunks/vidstack-CyIpzU-o.js';
import '../chunks/vidstack-D1JzjGR7.js';
import '../chunks/vidstack-awVFxJvn.js';
import '../chunks/vidstack-CerDQCFL.js';
import '../chunks/vidstack-B_9E7csz.js';
import '../chunks/vidstack-BYuWkQD6.js';
import '../chunks/vidstack-CMz4Bd8X.js';
import '../chunks/vidstack-B3EgHpRH.js';
import '../chunks/vidstack-2fIzRyOW.js';
import '../chunks/vidstack-XO9MYVWR.js';
import '../chunks/vidstack-ksPACRiU.js';

let DefaultAudioLayout$1 = class DefaultAudioLayout extends DefaultLayout {
  static {
    this.props = {
      ...super.props,
      when: ({ viewType }) => viewType === "audio",
      smallWhen: ({ width }) => width < 576
    };
  }
};

function DefaultAudioLayout() {
  return [
    x`<media-captions class="vds-captions"></media-captions>`,
    x`
      <media-controls class="vds-controls">
        <media-controls-group class="vds-controls-group">
          ${[
      DefaultSeekButton({ backward: true, tooltip: "top start" }),
      DefaultPlayButton({ tooltip: "top" }),
      DefaultSeekButton({ tooltip: "top" }),
      DefaultAudioTitle(),
      DefaultTimeSlider(),
      DefaultTimeInvert(),
      DefaultAudioVolume(),
      DefaultCaptionButton({ tooltip: "top" }),
      DefaultAudioMenus()
    ]}
        </media-controls-group>
      </media-controls>
    `
  ];
}
function DefaultAudioTitle() {
  return $signal(() => {
    let $ref = signal(void 0), $isTextOverflowing = signal(false), media = useMediaContext(), { title, started, currentTime, ended } = useMediaState(), { translations } = useDefaultLayoutContext(), $isTransitionActive = useTransitionActive($ref), $isContinued = () => started() || currentTime() > 0;
    const $title = () => {
      const word = ended() ? "Replay" : $isContinued() ? "Continue" : "Play";
      return `${i18n(translations, word)}: ${title()}`;
    };
    effect(() => {
      if ($isTransitionActive() && document.activeElement === document.body) {
        media.player.el?.focus();
      }
    });
    function onResize() {
      const el = $ref(), isOverflowing = !!el && !$isTransitionActive() && el.clientWidth < el.children[0].clientWidth;
      el && toggleClass(el, "vds-marquee", isOverflowing);
      $isTextOverflowing.set(isOverflowing);
    }
    function Title() {
      return x`
        <span class="vds-title-text">
          ${$signal($title)}${$signal(() => $isContinued() ? DefaultChapterTitle() : null)}
        </span>
      `;
    }
    useResizeObserver($ref, onResize);
    return title() ? x`
          <span class="vds-title" title=${$signal($title)} ${n($ref.set)}>
            ${[
      Title(),
      $signal(() => $isTextOverflowing() && !$isTransitionActive() ? Title() : null)
    ]}
          </span>
        ` : DefaultControlsSpacer();
  });
}
function DefaultAudioVolume() {
  return $signal(() => {
    const { pointer, muted } = useMediaState();
    if (pointer() === "coarse" && !muted())
      return null;
    const $rootRef = signal(void 0), $triggerRef = signal(void 0), $popperRef = signal(void 0), $isRootActive = useActive($rootRef), $hasMouseEnteredTrigger = useMouseEnter($triggerRef);
    effect(() => {
      if (!$hasMouseEnteredTrigger())
        return;
      useRectCSSVars($rootRef, $triggerRef, "trigger");
      useRectCSSVars($rootRef, $popperRef, "popper");
    });
    return x`
      <div class="vds-volume" ?data-active=${$signal($isRootActive)} ${n($rootRef.set)}>
        ${DefaultMuteButton({ tooltip: "top", ref: $triggerRef.set })}
        <div class="vds-volume-popup" ${n($popperRef.set)}>
          ${DefaultVolumeSlider({ orientation: "vertical" })}
        </div>
      </div>
    `;
  });
}
function DefaultAudioMenus() {
  const placement = "top end";
  return [
    DefaultChaptersMenu({ tooltip: "top", placement, portal: true }),
    DefaultSettingsMenu({ tooltip: "top end", placement, portal: true })
  ];
}

class MediaAudioLayoutElement extends Host(LitElement, DefaultAudioLayout$1) {
  constructor() {
    super(...arguments);
    this.jn = signal(false);
  }
  static {
    this.tagName = "media-audio-layout";
  }
  static {
    this.attrs = {
      smallWhen: {
        converter(value) {
          return value !== "never" && !!value;
        }
      }
    };
  }
  onSetup() {
    this.forwardKeepAlive = false;
    this.a = useMediaContext();
    this.classList.add("vds-audio-layout");
    this.menuContainer = createMenuContainer("vds-audio-layout");
    effect(() => {
      if (!this.menuContainer)
        return;
      setAttribute(this.menuContainer, "data-size", this.isSmallLayout && "sm");
    });
    const { pointer } = this.a.$state;
    effect(() => {
      if (pointer() !== "coarse")
        return;
      effect(this.kn.bind(this));
    });
    onDispose(() => this.menuContainer?.remove());
  }
  onConnect() {
    this.a.player.el?.setAttribute("data-layout", "audio");
    onDispose(() => this.a.player.el?.removeAttribute("data-layout"));
    effect(() => {
      if (this.$props.customIcons()) {
        new SlotManager(this).connect();
      } else {
        new DefaultLayoutIconsLoader(this).connect();
      }
    });
  }
  render() {
    return x`${$signal(this._j.bind(this))}`;
  }
  _j() {
    return this.isMatch ? DefaultAudioLayout() : null;
  }
  kn() {
    if (!this.jn()) {
      listenEvent(this, "pointerdown", this.ln.bind(this), { capture: true });
      return;
    }
    listenEvent(this, "pointerdown", (e) => e.stopPropagation());
    listenEvent(window, "pointerdown", this.mn.bind(this));
  }
  ln(event) {
    const { target } = event, hasTimeSlider = !!(target instanceof HTMLElement && target.closest(".vds-time-slider"));
    if (!hasTimeSlider)
      return;
    event.stopImmediatePropagation();
    this.setAttribute("data-scrubbing", "");
    this.jn.set(true);
  }
  mn() {
    this.jn.set(false);
    this.removeAttribute("data-scrubbing");
  }
}

export { MediaAudioLayoutElement };
