// feathers-utils/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldSkip = exports.onDelete = exports.getItemsIsArray = exports.filterQuery = exports.markHookForSkip = exports.setResultEmpty = exports.pushSet = exports.mergeArrays = exports.mergeQuery = exports.isPaginated = exports.isMulti = exports.getPaginate = exports.addHook = exports.DebouncedStore = exports.debounceMixin = exports.mixins = exports.runPerItem = exports.removeRelated = exports.setData = exports.createRelated = exports.checkMulti = exports.hooks = void 0;
// hooks
const checkMulti_1 = require("./hooks/checkMulti");
Object.defineProperty(exports, "checkMulti", { enumerable: true, get: function () { return checkMulti_1.checkMulti; } });
const setData_1 = require("./hooks/setData");
Object.defineProperty(exports, "setData", { enumerable: true, get: function () { return setData_1.setData; } });
const runPerItem_1 = require("./hooks/runPerItem");
Object.defineProperty(exports, "runPerItem", { enumerable: true, get: function () { return runPerItem_1.runPerItem; } });
exports.hooks = {
    checkMulti: checkMulti_1.checkMulti,
    setData: setData_1.setData,
    runPerItem: runPerItem_1.runPerItem,
};
var createRelated_1 = require("./hooks/createRelated");
Object.defineProperty(exports, "createRelated", { enumerable: true, get: function () { return createRelated_1.createRelated; } });
var removeRelated_1 = require("./hooks/removeRelated");
Object.defineProperty(exports, "removeRelated", { enumerable: true, get: function () { return removeRelated_1.removeRelated; } });
__exportStar(require("./hooks/forEach"), exports);
__exportStar(require("./hooks/parseFields"), exports);
const debounce_mixin_1 = require("./mixins/debounce-mixin");
Object.defineProperty(exports, "debounceMixin", { enumerable: true, get: function () { return debounce_mixin_1.debounceMixin; } });
Object.defineProperty(exports, "DebouncedStore", { enumerable: true, get: function () { return debounce_mixin_1.DebouncedStore; } });
exports.mixins = {
    debounceMixin: debounce_mixin_1.debounceMixin,
    DebouncedStore: debounce_mixin_1.DebouncedStore,
};
var addHook_1 = require("./utils/addHook");
Object.defineProperty(exports, "addHook", { enumerable: true, get: function () { return addHook_1.addHook; } });
var getPaginate_1 = require("./utils/getPaginate");
Object.defineProperty(exports, "getPaginate", { enumerable: true, get: function () { return getPaginate_1.getPaginate; } });
var isMulti_1 = require("./utils/isMulti");
Object.defineProperty(exports, "isMulti", { enumerable: true, get: function () { return isMulti_1.isMulti; } });
var isPaginated_1 = require("./utils/isPaginated");
Object.defineProperty(exports, "isPaginated", { enumerable: true, get: function () { return isPaginated_1.isPaginated; } });
var index_1 = require("./utils/mergeQuery/index");
Object.defineProperty(exports, "mergeQuery", { enumerable: true, get: function () { return index_1.mergeQuery; } });
var mergeArrays_1 = require("./utils/mergeQuery/mergeArrays");
Object.defineProperty(exports, "mergeArrays", { enumerable: true, get: function () { return mergeArrays_1.mergeArrays; } });
var pushSet_1 = require("./utils/pushSet");
Object.defineProperty(exports, "pushSet", { enumerable: true, get: function () { return pushSet_1.pushSet; } });
var setResultEmpty_1 = require("./utils/setResultEmpty");
Object.defineProperty(exports, "setResultEmpty", { enumerable: true, get: function () { return setResultEmpty_1.setResultEmpty; } });
var markHookForSkip_1 = require("./utils/markHookForSkip");
Object.defineProperty(exports, "markHookForSkip", { enumerable: true, get: function () { return markHookForSkip_1.markHookForSkip; } });
var filterQuery_1 = require("./utils/filterQuery");
Object.defineProperty(exports, "filterQuery", { enumerable: true, get: function () { return filterQuery_1.filterQuery; } });
var getItemsIsArray_1 = require("./utils/getItemsIsArray");
Object.defineProperty(exports, "getItemsIsArray", { enumerable: true, get: function () { return getItemsIsArray_1.getItemsIsArray; } });
var onDelete_1 = require("./hooks/onDelete");
Object.defineProperty(exports, "onDelete", { enumerable: true, get: function () { return onDelete_1.onDelete; } });
var shouldSkip_1 = require("./utils/shouldSkip");
Object.defineProperty(exports, "shouldSkip", { enumerable: true, get: function () { return shouldSkip_1.shouldSkip; } });
__exportStar(require("./types"), exports);
__exportStar(require("./utility-types/utility-types"), exports);
