// @marcellejs/backend/lib/services/assets/assets.hooks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const feathers_hooks_common_1 = require("feathers-hooks-common");
const rmdir_1 = __importDefault(require("../../utils/rmdir"));
const permission_hooks_1 = require("../../utils/permission-hooks");
// Don't remove this comment. It's needed to format import lines nicely.
const removeAssetFiles = async (context) => {
    if (!context.id) {
        return context;
    }
    const { url } = await context.service.get(context.id, { query: { $select: ['url'] } });
    const base = url.split('/')[0];
    (0, rmdir_1.default)(path_1.default.join(context.app.get('uploads'), 'assets', base));
    return context;
};
exports.default = (requireAuth) => {
    return {
        before: {
            all: [],
            find: [...(0, permission_hooks_1.authReadHooks)(requireAuth)],
            get: [...(0, permission_hooks_1.authReadHooks)(requireAuth)],
            create: [...(0, permission_hooks_1.authWriteHooks)(requireAuth), (0, feathers_hooks_common_1.setNow)('createdAt', 'updatedAt')],
            update: [...(0, permission_hooks_1.authWriteHooks)(requireAuth), removeAssetFiles, (0, feathers_hooks_common_1.setNow)('updatedAt')],
            patch: [
                ...(0, permission_hooks_1.authWriteHooks)(requireAuth),
                (0, feathers_hooks_common_1.iff)((context) => { var _a; return (_a = context.data) === null || _a === void 0 ? void 0 : _a.files; }, removeAssetFiles),
                (0, feathers_hooks_common_1.setNow)('updatedAt'),
            ],
            remove: [...(0, permission_hooks_1.authReadHooks)(requireAuth), removeAssetFiles],
        },
        after: {
            all: [],
            find: [],
            get: [],
            create: [],
            update: [],
            patch: [],
            remove: [],
        },
        error: {
            all: [],
            find: [],
            get: [],
            create: [],
            update: [],
            patch: [],
            remove: [],
        },
    };
};
