// vidstack/dist/prod/chunks/vidstack-CWofPlup.js
import { H as Host, e as effect, C as Component, n as signal, s as setAttribute, z as isBoolean, q as computed, t as isString } from './vidstack-DoOmgEfT.js';
import { ai as Captions, ah as Gesture, D as Controls, K as GoogleCastButton, I as ToggleButton, E as Tooltip, G as TooltipContent, a6 as MenuPortal, ab as ChaptersRadioGroup, aa as Radio, a9 as RadioGroup, S as Slider, _ as SliderVideo, a4 as SliderChapters } from './vidstack-D1JzjGR7.js';
import { u as useMediaContext } from './vidstack-i0oJKwBM.js';
import { C as ControlsGroup, T as TooltipTrigger } from './vidstack-BrgZdpmU.js';
import './vidstack-awVFxJvn.js';
import { e as watchCueTextChange } from './vidstack-CMz4Bd8X.js';
import { x } from './vidstack-Bf6opHxk.js';
import { r as requestScopedAnimationFrame, c as cloneTemplateContent, a as createTemplate, h as cloneTemplate } from './vidstack-CyIpzU-o.js';
import { L as LitElement } from './vidstack-CbhL8rpf.js';
import { x as renderMenuItemsTemplate } from './vidstack-DoG3FlyX.js';

class MediaCaptionsElement extends Host(HTMLElement, Captions) {
  static {
    this.tagName = "media-captions";
  }
}

class MediaGestureElement extends Host(HTMLElement, Gesture) {
  static {
    this.tagName = "media-gesture";
  }
}

class MediaControlsElement extends Host(HTMLElement, Controls) {
  static {
    this.tagName = "media-controls";
  }
}

class MediaControlsGroupElement extends Host(HTMLElement, ControlsGroup) {
  static {
    this.tagName = "media-controls-group";
  }
}

class Title extends Component {
}
class MediaTitleElement extends Host(HTMLElement, Title) {
  static {
    this.tagName = "media-title";
  }
  onSetup() {
    this.a = useMediaContext();
  }
  onConnect() {
    effect(this.uc.bind(this));
  }
  uc() {
    const { title } = this.a.$state;
    this.textContent = title();
  }
}

class ChapterTitle extends Component {
}
class MediaChapterTitleElement extends Host(HTMLElement, ChapterTitle) {
  static {
    this.tagName = "media-chapter-title";
  }
  onSetup() {
    this.a = useMediaContext();
    this.Xj = signal("");
  }
  onConnect() {
    const tracks = this.a.textTracks;
    watchCueTextChange(tracks, "chapters", this.Xj.set);
    effect(() => {
      this.textContent = this.Xj();
    });
  }
}

class Spinner extends Component {
  static {
    this.props = {
      size: 96,
      trackWidth: 8,
      fillPercent: 50
    };
  }
  onConnect(el) {
    requestScopedAnimationFrame(() => {
      if (!this.connectScope)
        return;
      const root = el.querySelector("svg"), track = root.firstElementChild, trackFill = track.nextElementSibling;
      effect(this.ea.bind(this, root, track, trackFill));
    });
  }
  ea(root, track, trackFill) {
    const { size, trackWidth, fillPercent } = this.$props;
    setAttribute(root, "width", size());
    setAttribute(root, "height", size());
    setAttribute(track, "stroke-width", trackWidth());
    setAttribute(trackFill, "stroke-width", trackWidth());
    setAttribute(trackFill, "stroke-dashoffset", 100 - fillPercent());
  }
}
class MediaSpinnerElement extends Host(LitElement, Spinner) {
  static {
    this.tagName = "media-spinner";
  }
  render() {
    return x`
      <svg fill="none" viewBox="0 0 120 120" aria-hidden="true" data-part="root">
        <circle cx="60" cy="60" r="54" stroke="currentColor" data-part="track"></circle>
        <circle
          cx="60"
          cy="60"
          r="54"
          stroke="currentColor"
          pathLength="100"
          stroke-dasharray="100"
          data-part="track-fill"
        ></circle>
      </svg>
    `;
  }
}

class MediaLayout extends Component {
  static {
    this.props = {
      when: false
    };
  }
}
class MediaLayoutElement extends Host(HTMLElement, MediaLayout) {
  static {
    this.tagName = "media-layout";
  }
  onSetup() {
    this.a = useMediaContext();
  }
  onConnect() {
    effect(this.Zm.bind(this));
  }
  Zm() {
    const root = this.firstElementChild, isTemplate = root?.localName === "template", isHTMLElement = root instanceof HTMLElement, when = this.$props.when(), matches = isBoolean(when) ? when : computed(() => when(this.a.player.state))();
    if (!matches) {
      if (isTemplate) {
        this.textContent = "";
        this.appendChild(root);
      } else if (isHTMLElement) {
        root.style.display = "none";
      }
      return;
    }
    if (isTemplate) {
      this.append(root.content.cloneNode(true));
    } else if (isHTMLElement) {
      root.style.display = "";
    }
  }
}

class MediaGoogleCastButtonElement extends Host(HTMLElement, GoogleCastButton) {
  static {
    this.tagName = "media-google-cast-button";
  }
}

class MediaToggleButtonElement extends Host(HTMLElement, ToggleButton) {
  static {
    this.tagName = "media-toggle-button";
  }
}

class MediaTooltipElement extends Host(HTMLElement, Tooltip) {
  static {
    this.tagName = "media-tooltip";
  }
}

class MediaTooltipTriggerElement extends Host(HTMLElement, TooltipTrigger) {
  static {
    this.tagName = "media-tooltip-trigger";
  }
  onConnect() {
    this.style.display = "contents";
  }
}

class MediaTooltipContentElement extends Host(HTMLElement, TooltipContent) {
  static {
    this.tagName = "media-tooltip-content";
  }
}

class MediaMenuPortalElement extends Host(HTMLElement, MenuPortal) {
  static {
    this.tagName = "media-menu-portal";
  }
  static {
    this.attrs = {
      disabled: {
        converter(value) {
          if (isString(value))
            return value;
          return value !== null;
        }
      }
    };
  }
}

class MediaChaptersRadioGroupElement extends Host(HTMLElement, ChaptersRadioGroup) {
  static {
    this.tagName = "media-chapters-radio-group";
  }
  onConnect() {
    renderMenuItemsTemplate(this, (el, option) => {
      const { cue, startTime, duration } = option, thumbnailEl = el.querySelector(".vds-thumbnail,media-thumbnail"), startEl = el.querySelector('[data-part="start-time"]'), durationEl = el.querySelector('[data-part="duration"]');
      if (startEl)
        startEl.textContent = startTime;
      if (durationEl)
        durationEl.textContent = duration;
      if (thumbnailEl) {
        thumbnailEl.setAttribute("time", cue.startTime + "");
        effect(() => {
          const thumbnails = this.$props.thumbnails();
          if ("src" in thumbnailEl) {
            thumbnailEl.src = thumbnails;
          } else if (isString(thumbnails)) {
            thumbnailEl.setAttribute("src", thumbnails);
          }
        });
      }
    });
  }
}

class MediaRadioElement extends Host(HTMLElement, Radio) {
  static {
    this.tagName = "media-radio";
  }
}

class MediaRadioGroupElement extends Host(HTMLElement, RadioGroup) {
  static {
    this.tagName = "media-radio-group";
  }
}

class MediaSliderElement extends Host(HTMLElement, Slider) {
  static {
    this.tagName = "media-slider";
  }
}

const videoTemplate = /* @__PURE__ */ createTemplate(
  `<video muted playsinline preload="none" style="max-width: unset;"></video>`
);
class MediaSliderVideoElement extends Host(HTMLElement, SliderVideo) {
  constructor() {
    super(...arguments);
    this.m = this.Yj();
  }
  static {
    this.tagName = "media-slider-video";
  }
  onSetup() {
    this.a = useMediaContext();
    this.$state.video.set(this.m);
  }
  onConnect() {
    const { canLoad } = this.a.$state, { src, crossOrigin } = this.$state;
    if (this.m.parentNode !== this) {
      this.prepend(this.m);
    }
    effect(() => {
      setAttribute(this.m, "crossorigin", crossOrigin());
      setAttribute(this.m, "preload", canLoad() ? "auto" : "none");
      setAttribute(this.m, "src", src());
    });
  }
  Yj() {
    return cloneTemplateContent(videoTemplate);
  }
}

class MediaSliderChaptersElement extends Host(HTMLElement, SliderChapters) {
  constructor() {
    super(...arguments);
    this.$j = null;
  }
  static {
    this.tagName = "media-slider-chapters";
  }
  onConnect() {
    requestScopedAnimationFrame(() => {
      if (!this.connectScope)
        return;
      const template = this.querySelector("template");
      if (template) {
        this.$j = template;
        effect(this.jk.bind(this));
      }
    });
  }
  jk() {
    if (!this.$j)
      return;
    const elements = cloneTemplate(this.$j, this.cues.length || 1);
    this.setRefs(elements);
  }
}

export { MediaCaptionsElement as M, MediaGestureElement as a, MediaControlsElement as b, MediaControlsGroupElement as c, MediaTitleElement as d, MediaChapterTitleElement as e, MediaSpinnerElement as f, MediaLayoutElement as g, MediaGoogleCastButtonElement as h, MediaToggleButtonElement as i, MediaTooltipElement as j, MediaTooltipTriggerElement as k, MediaTooltipContentElement as l, MediaMenuPortalElement as m, MediaChaptersRadioGroupElement as n, MediaRadioElement as o, MediaRadioGroupElement as p, MediaSliderElement as q, MediaSliderVideoElement as r, MediaSliderChaptersElement as s };
