// vidstack/dist/prod/chunks/vidstack-6eBVBXTk.js
import { c as coerceToError } from './vidstack-2fIzRyOW.js';
import { c as canGoogleCastSrc } from './vidstack-XO9MYVWR.js';
import { l as loadScript } from './vidstack-CMz4Bd8X.js';
import { I as IS_CHROME, a as IS_IOS } from './vidstack-B_9E7csz.js';
import { g as getCastContext, a as getCastSession, i as isCastConnected, h as hasLoadedCastFramework, b as isCastAvailable, c as getCastErrorMessage, d as getDefaultCastOptions, e as getCastFrameworkURL } from './vidstack-iUcBkPRm.js';
import { u as peek, D as DOMEvent } from './vidstack-DoOmgEfT.js';
import './vidstack-CyIpzU-o.js';

class GoogleCastLoader {
  constructor() {
    this.name = "google-cast";
  }
  /**
   * @see {@link https://developers.google.com/cast/docs/reference/web_sender/cast.framework.CastContext}
   */
  get cast() {
    return getCastContext();
  }
  mediaType() {
    return "video";
  }
  canPlay(src) {
    return IS_CHROME && !IS_IOS && canGoogleCastSrc(src);
  }
  async prompt(ctx) {
    let openEvent;
    try {
      const loadEvent = await this.Pl(ctx);
      if (!this.aa) {
        this.aa = new cast.framework.RemotePlayer();
        new cast.framework.RemotePlayerController(this.aa);
      }
      openEvent = ctx.player.createEvent("google-cast-prompt-open", {
        trigger: loadEvent
      });
      ctx.player.dispatchEvent(openEvent);
      this.pm(ctx, "connecting", openEvent);
      await this.Rl(peek(ctx.$props.googleCast));
      ctx.$state.remotePlaybackInfo.set({
        deviceName: getCastSession()?.getCastDevice().friendlyName
      });
      if (isCastConnected())
        this.pm(ctx, "connected", openEvent);
    } catch (message) {
      const error = coerceToError(message);
      this.pm(
        ctx,
        isCastConnected() ? "connected" : "disconnected",
        new DOMEvent("google-cast-prompt-error", { detail: error })
      );
      throw error;
    } finally {
      ctx.player.dispatch("google-cast-prompt-close", {
        trigger: openEvent
      });
    }
  }
  async load(ctx) {
    if (!this.aa) {
      throw Error("[vidstack] google cast player was not initialized");
    }
    return new (await import('../providers/vidstack-google-cast.js')).GoogleCastProvider(this.aa, ctx);
  }
  async Pl(ctx) {
    if (hasLoadedCastFramework())
      return;
    const loadStartEvent = ctx.player.createEvent("google-cast-load-start");
    ctx.player.dispatch(loadStartEvent);
    await loadScript(getCastFrameworkURL());
    await customElements.whenDefined("google-cast-launcher");
    const loadedEvent = ctx.player.createEvent("google-cast-loaded", { trigger: loadStartEvent });
    ctx.player.dispatch(loadedEvent);
    if (!isCastAvailable()) {
      throw Error(
        "Cast not available."
      );
    }
    return loadedEvent;
  }
  async Rl(options) {
    this.Tl(options);
    const errorCode = await this.cast.requestSession();
    if (errorCode)
      throw Error(getCastErrorMessage(errorCode));
  }
  Tl(options) {
    this.cast?.setOptions({
      ...getDefaultCastOptions(),
      ...options
    });
  }
  pm(ctx, state, trigger) {
    const detail = { type: "google-cast", state };
    ctx.delegate.c("remote-playback-change", detail, trigger);
  }
}

export { GoogleCastLoader };
