// vidstack/dist/prod/chunks/vidstack-BrgZdpmU.js
import { C as Component, j as setStyle, o as onDispose, v as useContext } from './vidstack-DoOmgEfT.js';
import { F as FocusVisibleController, t as tooltipContext } from './vidstack-D1JzjGR7.js';
import { r as requestScopedAnimationFrame } from './vidstack-CyIpzU-o.js';

class ControlsGroup extends Component {
  onAttach(el) {
    if (!el.style.pointerEvents)
      setStyle(el, "pointer-events", "auto");
  }
}

class TooltipTrigger extends Component {
  constructor() {
    super();
    new FocusVisibleController();
  }
  onConnect(el) {
    onDispose(
      requestScopedAnimationFrame(() => {
        if (!this.connectScope)
          return;
        this.Ja();
        const tooltip = useContext(tooltipContext);
        onDispose(() => {
          const button = this.Cc();
          button && tooltip.Kd(button);
        });
      })
    );
  }
  Ja() {
    const button = this.Cc(), tooltip = useContext(tooltipContext);
    button && tooltip.Jd(button);
  }
  Cc() {
    const candidate = this.el.firstElementChild;
    return candidate?.localName === "button" || candidate?.getAttribute("role") === "button" ? candidate : this.el;
  }
}

export { ControlsGroup as C, TooltipTrigger as T };
