// vidstack/dist/prod/chunks/vidstack-DPZGEOYG.js
import { A as deferredPromise, n as signal, l as listenEvent, e as effect, u as peek, t as isString } from './vidstack-DoOmgEfT.js';
import { a as appendParamsToURL } from './vidstack-CMz4Bd8X.js';

function timedPromise(callback, ms = 3e3) {
  const promise = deferredPromise();
  setTimeout(() => {
    const rejection = callback();
    if (rejection)
      promise.reject(rejection);
  }, ms);
  return promise;
}

class EmbedProvider {
  constructor(_iframe) {
    this.db = _iframe;
    this.cb = signal("");
    /**
     * Defines which referrer is sent when fetching the resource.
     *
     * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLIFrameElement/referrerPolicy}
     */
    this.referrerPolicy = null;
    _iframe.setAttribute("frameBorder", "0");
    _iframe.setAttribute("aria-hidden", "true");
    _iframe.setAttribute(
      "allow",
      "autoplay; fullscreen; encrypted-media; picture-in-picture; accelerometer; gyroscope"
    );
    if (this.referrerPolicy !== null) {
      _iframe.setAttribute("referrerpolicy", this.referrerPolicy);
    }
  }
  get iframe() {
    return this.db;
  }
  setup() {
    listenEvent(window, "message", this.Lg.bind(this));
    listenEvent(this.db, "load", this.lc.bind(this));
    effect(this.fd.bind(this));
  }
  fd() {
    const src = this.cb();
    if (!src.length) {
      this.db.setAttribute("src", "");
      return;
    }
    const params = peek(() => this.Te());
    this.db.setAttribute("src", appendParamsToURL(src, params));
  }
  gd(message, target) {
    this.db.contentWindow?.postMessage(JSON.stringify(message), target ?? "*");
  }
  Lg(event) {
    const origin = this.eb(), isOriginMatch = (event.source === null || event.source === this.db?.contentWindow) && (!isString(origin) || origin === event.origin);
    if (!isOriginMatch)
      return;
    try {
      const message = JSON.parse(event.data);
      if (message)
        this.hd(message, event);
      return;
    } catch (e) {
    }
    if (event.data)
      this.hd(event.data, event);
  }
}

export { EmbedProvider as E, timedPromise as t };
