// vidstack/dist/prod/chunks/vidstack-Vw-d0U2k.js
import { H as Host, e as effect, s as setAttribute, q as computed } from './vidstack-DoOmgEfT.js';
import { a as MediaProvider, b as MediaPlayer } from './vidstack-D1JzjGR7.js';
import { u as useMediaContext } from './vidstack-i0oJKwBM.js';

class MediaProviderElement extends Host(HTMLElement, MediaProvider) {
  constructor() {
    super(...arguments);
    this.A = null;
    this.Vj = null;
  }
  static {
    this.tagName = "media-provider";
  }
  onSetup() {
    this.a = useMediaContext();
    this.setAttribute("keep-alive", "");
  }
  onDestroy() {
    this.Vj?.remove();
    this.Vj = null;
    this.A?.remove();
    this.A = null;
  }
  onConnect() {
    effect(() => {
      const loader = this.$state.loader(), isYouTubeEmbed = loader?.name === "youtube", isVimeoEmbed = loader?.name === "vimeo", isEmbed = isYouTubeEmbed || isVimeoEmbed, isGoogleCast = loader?.name === "google-cast";
      const target = loader ? isGoogleCast ? this.Sm() : isEmbed ? this.ek() : loader.mediaType() === "audio" ? this.fk() : this.Yj() : null;
      if (this.A !== target) {
        const parent = this.A?.parentElement ?? this;
        this.A?.remove();
        this.A = target;
        if (target)
          parent.prepend(target);
        if (isEmbed && target) {
          effect(() => {
            const { $iosControls } = this.a, { controls } = this.a.$state, showControls = controls() || $iosControls();
            if (showControls) {
              this.Vj?.remove();
              this.Vj = null;
            } else {
              this.Vj = this.querySelector(".vds-blocker") ?? document.createElement("div");
              this.Vj.classList.add("vds-blocker");
              target.after(this.Vj);
            }
            setAttribute(target, "data-no-controls", !showControls);
          });
        }
      }
      if (isYouTubeEmbed)
        target?.classList.add("vds-youtube");
      else if (isVimeoEmbed)
        target?.classList.add("vds-vimeo");
      if (!isEmbed) {
        this.Vj?.remove();
        this.Vj = null;
      }
      this.load(target);
    });
  }
  fk() {
    const audio = this.A instanceof HTMLAudioElement ? this.A : document.createElement("audio");
    setAttribute(audio, "preload", "none");
    setAttribute(audio, "aria-hidden", "true");
    const { controls, crossOrigin } = this.a.$state;
    effect(() => {
      setAttribute(audio, "controls", controls());
      setAttribute(audio, "crossorigin", crossOrigin());
    });
    return audio;
  }
  Yj() {
    const video = this.A instanceof HTMLVideoElement ? this.A : document.createElement("video");
    const { controls, crossOrigin, poster } = this.a.$state, { $iosControls } = this.a, $nativeControls = computed(() => controls() || $iosControls() ? "true" : null), $poster = computed(() => poster() && (controls() || $iosControls()) ? poster() : null);
    effect(() => {
      setAttribute(video, "controls", $nativeControls());
      setAttribute(video, "crossorigin", crossOrigin());
      setAttribute(video, "poster", $poster());
    });
    return video;
  }
  ek() {
    const iframe = this.A instanceof HTMLIFrameElement ? this.A : document.createElement("iframe");
    const { controls } = this.a.$state, { $iosControls } = this.a, $nativeControls = computed(() => controls() || $iosControls());
    effect(() => setAttribute(iframe, "tabindex", !$nativeControls() ? -1 : null));
    return iframe;
  }
  Sm() {
    if (this.A?.classList.contains("vds-google-cast")) {
      return this.A;
    }
    const container = document.createElement("div");
    container.classList.add("vds-google-cast");
    import('./vidstack-DqoEU8C6.js').then(({ insertContent }) => {
      insertContent(container, this.a.$state);
    });
    return container;
  }
}

class MediaPlayerElement extends Host(HTMLElement, MediaPlayer) {
  static {
    this.tagName = "media-player";
  }
  static {
    this.attrs = {
      autoPlay: "autoplay",
      crossOrigin: "crossorigin",
      playsInline: "playsinline",
      preferNativeHLS: "prefer-native-hls",
      minLiveDVRWindow: "min-live-dvr-window"
    };
  }
}

export { MediaProviderElement as M, MediaPlayerElement as a };
