// vidstack/dist/prod/providers/vidstack-video.js
import { o as onDispose, l as listenEvent, D as DOMEvent, n as signal, e as effect, h as scoped } from '../chunks/vidstack-DoOmgEfT.js';
import { e as canUsePictureInPicture, f as canUseVideoPresentation, c as canPlayHLSNatively } from '../chunks/vidstack-B_9E7csz.js';
import { HTMLMediaProvider } from './vidstack-html.js';
import { T as TextTrack, b as TextTrackSymbol } from '../chunks/vidstack-CMz4Bd8X.js';
import '../chunks/vidstack-XO9MYVWR.js';
import '../chunks/vidstack-BhYx9Fjk.js';
import '../chunks/vidstack-ksPACRiU.js';
import '../chunks/vidstack-CerDQCFL.js';

class NativeHLSTextTracks {
  constructor(_video, _ctx) {
    this.m = _video;
    this.b = _ctx;
    _video.textTracks.onaddtrack = this.Wc.bind(this);
    onDispose(this.cd.bind(this));
  }
  Wc(event) {
    const nativeTrack = event.track;
    if (!nativeTrack || findTextTrackElement(this.m, nativeTrack))
      return;
    const track = new TextTrack({
      id: nativeTrack.id,
      kind: nativeTrack.kind,
      label: nativeTrack.label,
      language: nativeTrack.language,
      type: "vtt"
    });
    track[TextTrackSymbol.T] = { track: nativeTrack };
    track[TextTrackSymbol.M] = 2;
    track[TextTrackSymbol.te] = true;
    let lastIndex = 0;
    const onCueChange = (event2) => {
      if (!nativeTrack.cues)
        return;
      for (let i = lastIndex; i < nativeTrack.cues.length; i++) {
        track.addCue(nativeTrack.cues[i], event2);
        lastIndex++;
      }
    };
    onCueChange(event);
    nativeTrack.oncuechange = onCueChange;
    this.b.textTracks.add(track, event);
    track.setMode(nativeTrack.mode, event);
  }
  cd() {
    this.m.textTracks.onaddtrack = null;
    for (const track of this.b.textTracks) {
      const nativeTrack = track[TextTrackSymbol.T]?.track;
      if (nativeTrack?.oncuechange)
        nativeTrack.oncuechange = null;
    }
  }
}
function findTextTrackElement(video, track) {
  return Array.from(video.children).find((el) => el.track === track);
}

class VideoPictureInPicture {
  constructor(_video, _media) {
    this.m = _video;
    this.a = _media;
    this.B = (active, event) => {
      this.a.delegate.c("picture-in-picture-change", active, event);
    };
    listenEvent(this.m, "enterpictureinpicture", this.Jg.bind(this));
    listenEvent(this.m, "leavepictureinpicture", this.Kg.bind(this));
  }
  get active() {
    return document.pictureInPictureElement === this.m;
  }
  get supported() {
    return canUsePictureInPicture(this.m);
  }
  async enter() {
    return this.m.requestPictureInPicture();
  }
  exit() {
    return document.exitPictureInPicture();
  }
  Jg(event) {
    this.B(true, event);
  }
  Kg(event) {
    this.B(false, event);
  }
}

class VideoPresentation {
  constructor(_video, _media) {
    this.m = _video;
    this.a = _media;
    this.I = "inline";
    listenEvent(this.m, "webkitpresentationmodechanged", this.Ua.bind(this));
  }
  get Se() {
    return canUseVideoPresentation(this.m);
  }
  async kc(mode) {
    if (this.I === mode)
      return;
    this.m.webkitSetPresentationMode(mode);
  }
  Ua(event) {
    const prevMode = this.I;
    this.I = this.m.webkitPresentationMode;
    this.a.player?.dispatch(
      new DOMEvent("video-presentation-change", {
        detail: this.I,
        trigger: event
      })
    );
    ["fullscreen", "picture-in-picture"].forEach((type) => {
      if (this.I === type || prevMode === type) {
        this.a.delegate.c(`${type}-change`, this.I === type, event);
      }
    });
  }
}
class FullscreenPresentationAdapter {
  constructor(_presentation) {
    this.fa = _presentation;
  }
  get active() {
    return this.fa.I === "fullscreen";
  }
  get supported() {
    return this.fa.Se;
  }
  async enter() {
    this.fa.kc("fullscreen");
  }
  async exit() {
    this.fa.kc("inline");
  }
}
class PIPPresentationAdapter {
  constructor(_presentation) {
    this.fa = _presentation;
  }
  get active() {
    return this.fa.I === "picture-in-picture";
  }
  get supported() {
    return this.fa.Se;
  }
  async enter() {
    this.fa.kc("picture-in-picture");
  }
  async exit() {
    this.fa.kc("inline");
  }
}

class VideoRemotePlaybackAdapter {
  constructor(_video, _ctx) {
    this.m = _video;
    this.b = _ctx;
    this.Se = signal(false);
    this.Hm();
  }
  get supported() {
    return this.Se();
  }
  Hm() {
    if (!this.m.remote || !this.sl)
      return;
    this.m.remote.watchAvailability((available) => {
      this.Se.set(available);
    }).catch(() => {
      this.Se.set(false);
    });
    effect(this.rn.bind(this));
  }
  rn() {
    if (!this.Se())
      return;
    const events = ["connecting", "connect", "disconnect"], onStateChange = this.Rg.bind(this);
    onStateChange();
    listenEvent(this.m, "playing", onStateChange);
    for (const type of events) {
      listenEvent(this.m.remote, type, onStateChange);
    }
  }
  async prompt() {
    if (!this.supported)
      throw Error("Not supported on this platform.");
    if (this.vb === "airplay" && this.m.webkitShowPlaybackTargetPicker) {
      return this.m.webkitShowPlaybackTargetPicker();
    }
    return this.m.remote.prompt();
  }
  Rg(event) {
    const state = this.m.remote.state;
    if (state === this.mc)
      return;
    const detail = { type: this.vb, state };
    this.b.delegate.c("remote-playback-change", detail, event);
    this.mc = state;
  }
}
class VideoAirPlayAdapter extends VideoRemotePlaybackAdapter {
  constructor() {
    super(...arguments);
    this.vb = "airplay";
  }
  get sl() {
    return "WebKitPlaybackTargetAvailabilityEvent" in window;
  }
}

class VideoProvider extends HTMLMediaProvider {
  constructor(video, ctx) {
    super(video, ctx);
    this.$$PROVIDER_TYPE = "VIDEO";
    scoped(() => {
      this.airPlay = new VideoAirPlayAdapter(video, ctx);
      if (canUseVideoPresentation(video)) {
        const presentation = new VideoPresentation(video, ctx);
        this.fullscreen = new FullscreenPresentationAdapter(presentation);
        this.pictureInPicture = new PIPPresentationAdapter(presentation);
      } else if (canUsePictureInPicture(video)) {
        this.pictureInPicture = new VideoPictureInPicture(video, ctx);
      }
    }, this.scope);
  }
  get type() {
    return "video";
  }
  setup() {
    super.setup();
    if (canPlayHLSNatively(this.video)) {
      new NativeHLSTextTracks(this.video, this.b);
    }
    this.b.textRenderers.Fe(this.video);
    onDispose(() => {
      this.b.textRenderers.Fe(null);
    });
    if (this.type === "video")
      this.b.delegate.c("provider-setup", this);
  }
  /**
   * The native HTML `<video>` element.
   *
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement}
   */
  get video() {
    return this.a;
  }
}

export { VideoProvider };
